web/packages/shared/hooks/useInfiniteScroll/useInfiniteScroll.ts

```diff
@@ -14,32 +14,24 @@
  * limitations under the License.
  */
 
-import { useLayoutEffect, useRef } from 'react';
-
-import { Attempt } from 'shared/hooks/useAttemptNext';
-
-import {
-  useKeyBasedPagination,
-  Props as PaginationProps,
-} from './useKeyBasedPagination';
+import { useLayoutEffect, useRef, useCallback } from 'react';
 
 /**
- * Fetches a part of resource list whenever the `trigger` element intersects the
+ * Calls fetch function whenever the `trigger` element intersects the
  * viewport until the list is exhausted or an error happens.
  *
  * Callers must set the `trigger` element by passing the [`State.setTrigger`] function
  * as the `ref` prop of the element they want to use as the trigger.
- *
- * Use the [`State.forceFetch`] to continue after an error.
  */
-export function useInfiniteScroll<T>(props: PaginationProps<T>): State<T> {
+export function useInfiniteScroll({
+  fetch,
+}: {
+  fetch: () => Promise<void>;
+}): InfiniteScroll {
   const observer = useRef<IntersectionObserver | null>(null);
   const trigger = useRef<Element | null>(null);
 
-  const { fetch, forceFetch, attempt, resources } =
-    useKeyBasedPagination(props);
-
-  const recreateObserver = () => {
+  const recreateObserver = useCallback(() => {
```

Comment:
Not particularly relevant to this PR itself, but when I was trying to understand this code in general (which was easier thanks to comments Bartosz and Grzegorz wrote), one thing that stuck out to me is that I couldn't easily identify when the initial fetch happens and what causes it.

But I understand the assumption is that at the start, the trigger element is within the viewport and thus `useInfiniteScroll` triggers the first fetch, right?

Reply:
Yeah, that's correct, `useInfiniteScroll` triggers the initial request too.

Reply:
Yes. For me, it was obvious, but perhaps worth documenting, too.

Reply:
Added a comment.